File: PPI-Backend/app/routers/auditoria/auditoria.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Dict, Optional
from datetime import datetime
from app.database.connection import get_db
from app.services.auditoria.auditoria_service import AuditoriaService
from app.core.auth.dependencies import get_current_admin_user

router = APIRouter(
    prefix="/auditoria",
    tags=["auditoria"]
)


def get_auditoria_service(db: AsyncSession = Depends(get_db)) -> AuditoriaService:
    """Dependency para obtener el servicio de auditoría"""
    return AuditoriaService(db)


@router.get(
    "/dashboard",
    summary="Obtener dashboard completo",
    description="Obtiene todos los datos del dashboard de auditoría incluyendo KPIs, tendencias, rankings y distribución"
)
async def get_dashboard(
    evento_id: Optional[int] = Query(None, description="ID del evento específico", ge=1),
    fecha_desde: Optional[datetime] = Query(None, description="Fecha de inicio del rango (YYYY-MM-DD o YYYY-MM-DDTHH:MM:SS)"),
    fecha_hasta: Optional[datetime] = Query(None, description="Fecha de fin del rango (YYYY-MM-DD o YYYY-MM-DDTHH:MM:SS)"),
    current_user = Depends(get_current_admin_user),
    auditoria_service: AuditoriaService = Depends(get_auditoria_service)
) -> Dict:
    """
    Obtiene todos los datos del dashboard de auditoría
    
    **Parámetros:**
    - `evento_id`: (Opcional) Filtrar por evento específico
    - `fecha_desde`: (Opcional) Fecha de inicio. Por defecto: 1 año atrás
    - `fecha_hasta`: (Opcional) Fecha de fin. Por defecto: hoy
    
    **Retorna:**
    - KPIs principales
    - Datos mensuales para gráficos de tendencia
    - Top 5 eventos por ventas
    - Top 5 usuarios por compras
    - Top 5 locales por ingresos
    - Distribución de ventas por evento
    """
    return await auditoria_service.get_all_dashboard(evento_id, fecha_desde, fecha_hasta)


@router.get(
    "/kpis",
    summary="Obtener KPIs principales",
    description="Obtiene los indicadores clave de rendimiento del dashboard"
)
async def get_kpis(
    evento_id: Optional[int] = Query(None, description="ID del evento específico", ge=1),
    fecha_desde: Optional[datetime] = Query(None, description="Fecha de inicio del rango"),
    fecha_hasta: Optional[datetime] = Query(None, description="Fecha de fin del rango"),
    current_user = Depends(get_current_admin_user),
    auditoria_service: AuditoriaService = Depends(get_auditoria_service)
) -> Dict:
    """
    Obtiene solo los KPIs principales
    
    **KPIs retornados:**
    - Ventas totales
    - Ventas estimadas (precio de lista)
    - Tickets emitidos
    - Tickets transferidos
    - Incidencias
    - Tasa de conversión
    """
    return await auditoria_service.get_kpis(evento_id, fecha_desde, fecha_hasta)


@router.get(
    "/tendencias",
    summary="Obtener tendencias mensuales",
    description="Obtiene datos agregados por mes para gráficos de tendencia"
)
async def get_tendencias(
    evento_id: Optional[int] = Query(None, description="ID del evento específico", ge=1),
    fecha_desde: Optional[datetime] = Query(None, description="Fecha de inicio del rango"),
    fecha_hasta: Optional[datetime] = Query(None, description="Fecha de fin del rango"),
    current_user = Depends(get_current_admin_user),
    auditoria_service: AuditoriaService = Depends(get_auditoria_service)
) -> Dict:
    """
    Obtiene datos mensuales para gráficos de tendencia
    
    **Datos por mes:**
    - Ventas
    - Ventas estimadas
    - Tickets emitidos
    - Incidencias
    """
    return await auditoria_service.get_tendencias(evento_id, fecha_desde, fecha_hasta)


@router.get(
    "/rankings",
    summary="Obtener rankings",
    description="Obtiene los top eventos, usuarios y locales"
)
async def get_rankings(
    fecha_desde: Optional[datetime] = Query(None, description="Fecha de inicio del rango"),
    fecha_hasta: Optional[datetime] = Query(None, description="Fecha de fin del rango"),
    limite: int = Query(10, description="Límite de resultados", ge=1, le=100),
    current_user = Depends(get_current_admin_user),
    auditoria_service: AuditoriaService = Depends(get_auditoria_service)
) -> Dict:
    """
    Obtiene los rankings (top N)
    
    **Rankings retornados:**
    - Top eventos por ventas
    - Top usuarios por compras
    - Top locales por ingresos
    """
    return await auditoria_service.get_rankings(fecha_desde, fecha_hasta, limite)


@router.get(
    "/detalle",
    summary="Obtener detalle de transacciones",
    description="Obtiene el detalle completo de todas las transacciones"
)
async def get_detalle_transacciones(
    evento_id: Optional[int] = Query(None, description="ID del evento específico", ge=1),
    fecha_desde: Optional[datetime] = Query(None, description="Fecha de inicio del rango"),
    fecha_hasta: Optional[datetime] = Query(None, description="Fecha de fin del rango"),
    current_user = Depends(get_current_admin_user),
    auditoria_service: AuditoriaService = Depends(get_auditoria_service)
) -> Dict:
    """
    Obtiene el detalle de transacciones
    
    **Información por transacción:**
    - ID de pago
    - Cliente
    - Evento
    - Fecha y hora
    - Total
    - Método de pago
    - Cantidad de tickets
    """
    return await auditoria_service.get_detalle_transacciones(evento_id, fecha_desde, fecha_hasta)


@router.get(
    "/health",
    summary="Health check",
    description="Verifica el estado del servicio de auditoría"
)
async def health_check() -> Dict:
    """
    Health check del servicio de auditoría
    """
    return {
        "success": True,
        "message": "Servicio de auditoría funcionando correctamente",
        "data": {"status": "healthy"}
    }


@router.get(
    "/exportar",
    summary="Obtener datos completos para exportación",
    description="Obtiene TODOS los datos sin límites para exportar a Excel"
)
async def get_datos_exportacion(
    evento_id: Optional[int] = Query(None, description="ID del evento específico", ge=1),
    fecha_desde: Optional[datetime] = Query(None, description="Fecha de inicio del rango"),
    fecha_hasta: Optional[datetime] = Query(None, description="Fecha de fin del rango"),
    current_user = Depends(get_current_admin_user),
    auditoria_service: AuditoriaService = Depends(get_auditoria_service)
) -> Dict:
    """
    Obtiene todos los datos para exportación a Excel
    
    **Datos retornados (sin límites):**
    - KPIs principales
    - Datos mensuales completos
    - TODOS los eventos con ventas
    - TODOS los usuarios con compras
    - TODOS los locales con ingresos
    - Distribución por categorías
    - Detalle de TODAS las transacciones
    """
    return await auditoria_service.get_datos_exportacion(evento_id, fecha_desde, fecha_hasta)

File: PPI-Frontend/lib/api/auditoria.ts
// lib/api/auditoria.ts
// API Client para consumir los endpoints de auditoría

import { tokenUtils } from '@/lib/auth/token';
import type {
  LogErrorFilters,
  LogErrorListResponse,
  LogExportFormat,
  LogExportParams,
  LogExportResult,
} from '@/lib/types/auditoria/logs';

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';
const AUDITORIA_BASE_URL = `${API_BASE_URL}/auditoria`;

/**
 * Interfaces para los tipos de datos retornados por la API
 */

export interface KPIDashboard {
  ventas_totales: number;
  ventas_estimadas: number;
  tickets_emitidos: number;
  tickets_transferidos: number;
  incidencias: number;
  tasa_conversion: number;
  aforo_vendido: number;
  velocidad_venta: number;
}

export interface DatoMensual {
  mes: string;
  mes_nombre: string;
  ventas: number;
  ventas_estimadas: number;
  tickets: number;
  incidencias: number;
}

export interface TopEvento {
  id: number;
  nombre: string;
  ventas: number;
  tickets: number;
}

export interface TopUsuario {
  id: number;
  nombre_completo: string;
  email: string;
  total_compras: number;
  cantidad_compras: number;
  rango_nombre: string;
  puntos_disponibles: number;
}

export interface TopLocal {
  id: number;
  nombre: string;
  direccion: string;
  total_ingresos: number;
  cantidad_eventos: number;
}

export interface DistribucionCategoria {
  categoria_id: number;
  categoria_nombre: string;
  total_ventas: number;
  cantidad_eventos: number;
  porcentaje: number;
}

export interface DetalleTransaccion {
  pago_id: number;
  cliente_id: number;
  cliente_nombre: string;
  evento_id: number;
  evento_nombre: string;
  fecha_transaccion: string;
  total: number;
  metodo_pago: string;
  cantidad_tickets: number;
}

export interface DashboardCompleto {
  kpis: KPIDashboard;
  datos_mensuales: DatoMensual[];
  top_eventos: TopEvento[];
  top_usuarios: TopUsuario[];
  top_locales: TopLocal[];
  distribucion_categorias: DistribucionCategoria[];
}

export interface DatosExportacion {
  kpis: KPIDashboard;
  datos_mensuales: DatoMensual[];
  eventos: TopEvento[];
  usuarios: TopUsuario[];
  locales: TopLocal[];
  categorias: DistribucionCategoria[];
  transacciones: DetalleTransaccion[];
}

export interface ApiResponse<T> {
  success: boolean;
  message: string;
  data: T | null;
}

/**
 * Parámetros de filtro opcionales
 */
export interface FiltrosReporteria {
  evento_id?: number;
  fecha_desde?: string; // ISO 8601 format
  fecha_hasta?: string; // ISO 8601 format
  limite?: number;
}

/**
 * Función helper para hacer requests a la API
 */
async function apiRequest<T>(
  endpoint: string,
  params?: FiltrosReporteria
): Promise<ApiResponse<T>> {
  const token = await tokenUtils.getToken();
  
  if (!token) {
    throw new Error('No se encontró token de autenticación');
  }

  // Construir query params
  const queryParams = new URLSearchParams();
  if (params) {
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined && value !== null) {
        queryParams.append(key, value.toString());
      }
    });
  }

  const url = `${AUDITORIA_BASE_URL}/${endpoint}${
    queryParams.toString() ? `?${queryParams}` : ''
  }`;

  const response = await fetch(url, {
    method: 'GET',
    headers: {
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json',
      'Accept': 'application/json',
    },
    cache: 'no-store', // Siempre obtener datos frescos
  });

  if (!response.ok) {
    if (response.status === 401) {
      throw new Error('No autorizado. Por favor inicia sesión nuevamente.');
    }
    if (response.status === 403) {
      throw new Error('No tienes permisos para acceder a este recurso.');
    }
    throw new Error(`Error ${response.status}: ${response.statusText}`);
  }

  return response.json();
}

/**
 * API Functions
 */

/**
 * Obtiene todos los datos del dashboard en una sola llamada
 */
export async function getDashboardCompleto(
  filtros?: FiltrosReporteria
): Promise<ApiResponse<DashboardCompleto>> {
  return apiRequest<DashboardCompleto>('dashboard', filtros);
}

/**
 * Obtiene solo los KPIs principales
 */
export async function getKPIs(
  filtros?: FiltrosReporteria
): Promise<ApiResponse<KPIDashboard>> {
  return apiRequest<KPIDashboard>('kpis', filtros);
}

/**
 * Obtiene datos mensuales para gráficos de tendencia
 */
export async function getTendencias(
  filtros?: FiltrosReporteria
): Promise<ApiResponse<DatoMensual[]>> {
  return apiRequest<DatoMensual[]>('tendencias', filtros);
}

/**
 * Obtiene los rankings (top eventos, usuarios y locales)
 */
export async function getRankings(
  filtros?: Omit<FiltrosReporteria, 'evento_id'>
): Promise<ApiResponse<{
  top_eventos: TopEvento[];
  top_usuarios: TopUsuario[];
  top_locales: TopLocal[];
}>> {
  return apiRequest('rankings', filtros);
}

/**
 * Obtiene el detalle de transacciones
 */
export async function getDetalleTransacciones(
  filtros?: FiltrosReporteria
): Promise<ApiResponse<DetalleTransaccion[]>> {
  return apiRequest('detalle', filtros);
}

/**
 * Obtiene TODOS los datos para exportación a Excel (sin límites)
 */
export async function getDatosExportacion(
  filtros?: FiltrosReporteria
): Promise<ApiResponse<DatosExportacion>> {
  return apiRequest<DatosExportacion>('exportar', filtros);
}

/**
 * Health check del servicio
 */
export async function healthCheck(): Promise<ApiResponse<{ status: string }>> {
  const response = await fetch(`${AUDITORIA_BASE_URL}/health`, {
    method: 'GET',
    headers: {
      'Content-Type': 'application/json',
    },
  });

  if (!response.ok) {
    throw new Error('Servicio de auditoría no disponible');
  }

  return response.json();
}

/**
 * Formatea una fecha a ISO 8601 para la API
 */
export function formatearFechaParaAPI(fecha: Date): string {
  return fecha.toISOString();
}

/**
 * Obtiene el rango de fechas de los últimos N meses
 */
export function obtenerRangoUltimosMeses(meses: number): {
  fecha_desde: string;
  fecha_hasta: string;
} {
  const fecha_hasta = new Date();
  const fecha_desde = new Date();
  
  if (meses === 1) {
    // Para "último mes", usar desde el día 1 del mes actual
    fecha_desde.setDate(1);
    fecha_desde.setHours(0, 0, 0, 0);
  } else {
    // Para múltiples meses, restar N meses completos
    fecha_desde.setMonth(fecha_desde.getMonth() - meses + 1);
    fecha_desde.setDate(1);
    fecha_desde.setHours(0, 0, 0, 0);
  }

  return {
    fecha_desde: formatearFechaParaAPI(fecha_desde),
    fecha_hasta: formatearFechaParaAPI(fecha_hasta),
  };
}

const buildQueryString = (params?: Record<string, string | number | undefined | null>) => {
  const query = new URLSearchParams();
  if (!params) {
    return '';
  }

  Object.entries(params).forEach(([key, value]) => {
    if (value === undefined || value === null || value === '') {
      return;
    }
    query.append(key, String(value));
  });

  const queryString = query.toString();
  return queryString ? `?${queryString}` : '';
};

const ensureBrowserToken = async () => {
  if (typeof window === 'undefined') {
    throw new Error('La consulta de logs solo está disponible desde el cliente');
  }

  const token = await tokenUtils.getToken();
  if (!token) {
    throw new Error('Sesión no válida. Inicia sesión nuevamente.');
  }
  return token;
};

export async function getLogErrors(
  filtros: LogErrorFilters
): Promise<LogErrorListResponse> {
  const token = await ensureBrowserToken();
  const queryString = buildQueryString(filtros as Record<string, string | number | undefined>);

  const response = await fetch(`${AUDITORIA_BASE_URL}/logs${queryString}`, {
    method: 'GET',
    headers: {
      Authorization: `Bearer ${token}`,
      Accept: 'application/json',
    },
    cache: 'no-store',
  });

  if (!response.ok) {
    const errorPayload = await response.json().catch(() => ({ detail: response.statusText }));
    throw new Error(errorPayload.detail || 'No se pudieron obtener los logs');
  }

  return response.json() as Promise<LogErrorListResponse>;
}

const extractFilename = (header: string | null, fallback: string) => {
  if (!header) {
    return fallback;
  }

  const match = header.match(/filename\*=UTF-8''(.+)|filename="?([^";]+)"?/i);
  if (match) {
    return decodeURIComponent(match[1] || match[2]);
  }
  return fallback;
};

export async function exportLogErrors(
  filtros: LogExportParams
): Promise<LogExportResult> {
  const token = await ensureBrowserToken();
  const queryString = buildQueryString(filtros as Record<string, string | number | undefined>);

  const response = await fetch(`${AUDITORIA_BASE_URL}/logs/export${queryString}`, {
    method: 'GET',
    headers: {
      Authorization: `Bearer ${token}`,
    },
  });

  if (!response.ok) {
    const errorPayload = await response.json().catch(() => ({ detail: response.statusText }));
    throw new Error(errorPayload.detail || 'No se pudo exportar el archivo');
  }

  const blob = await response.blob();
  const mediaType = response.headers.get('content-type') || 'application/octet-stream';
  const fallbackName = `logs-errores.${(filtros.formato || 'csv') as LogExportFormat}`;
  const filename = extractFilename(response.headers.get('content-disposition'), fallbackName);

  return { blob, filename, mediaType };
}